QUIC RADIX 6D Test Framework

hlandau opened the pull request:
Built on the server API base implementation PR.

Successor test framework to `quic_multistream_test`, which will be needed for testing the server API properly.

Long term plan is to get rid of the multistream test in favour of RADIX and remove `QUIC_TSERVER` &mdash; but there are a lot of tests that rely on both of these and that will have to wait.

See README.md for design discussion.

```
77737beb26 test/build.info: Ensure QUIC tests are properly disabled in no-quic builds
291186a926 QUIC RADIX: Add build configuration
6b39377b33 QUIC RADIX: Add recipe
c15c51ac38 testutil: Allow overriding output BIOs thread-locally
52ed6a5730 QUIC RADIX: Add RADIX test framework implementation
e5b4fd93fc QUIC RADIX: Add README
1279a3b14d QUIC RADIX: Add test of multithreading functionality
```

Amongst other things, this test framework is intended to offer better diagnostic output than `quic_multistream_test`. Here's an example showing a simple connection test and a trivial multithread test:

```
1..1
    # Subtest: test_script
    1..2
    # INFO:  @ test/radix/terp.c:525
    # generating script: script 'simple_conn' (simple connection to server)
    # 
    # Generated script for 'simple_conn':
    # 
    # --GENERATED-----------------------------------------------------------
    #   # NAME:
    #   #   simple_conn
    #   # SOURCE:
    #   #   test/radix/quic_tests.c:15
    #   # DESCRIPTION:
    #   #   simple connection to server
    # 
    #   0-     0>	PUSH_P   	           0x1207fe0 (L)
    #   1-    10>	PUSH_U64 	                 0x1 (1)
    #   2-    20>	FUNC     	hf_new_ssl
    #   3-    38>	PUSH_U64 	                 0x0 (0)
    #   4-    48>	PUSH_P   	           0x1207fe0 (L)
    #   5-    58>	FUNC     	hf_select_ssl
    #   6-    70>	FUNC     	hf_listen
    #   7-    88>	PUSH_P   	           0x12080e0 (C)
    #   8-    98>	PUSH_U64 	                 0x0 (0)
    #   9-    a8>	FUNC     	hf_new_ssl
    #  10-    c0>	PUSH_U64 	                 0x0 (0)
    #  11-    d0>	PUSH_P   	           0x12080e0 (C)
    #  12-    e0>	FUNC     	hf_select_ssl
    #  13-    f8>	PUSH_U64 	                 0x1 (1)
    #  14-   108>	PUSH_P   	           0x1207fe0 (L)
    #  15-   118>	FUNC     	hf_select_ssl
    #  16-   130>	FUNC     	hf_set_peer_addr_from
    #  17-   148>	PUSH_U64 	                 0x0 (0)
    #  18-   158>	PUSH_P   	           0x12080e0 (C)
    #  19-   168>	FUNC     	hf_select_ssl
    #  20-   180>	FUNC     	hf_connect_wait
    #  21-   198>	PUSH_U64 	                 0x0 (0)
    #  22-   1a8>	PUSH_P   	           0x12080e0 (C)
    #  23-   1b8>	FUNC     	hf_select_ssl
    #  24-   1d0>	PUSH_P   	           0x1206a60
    #  25-   1e0>	PUSH_SIZE 	                 0x6 (6)
    #  26-   1f0>	FUNC     	hf_write
    #  27-   208>	PUSH_U64 	                 0x0 (0)
    #  28-   218>	PUSH_P   	           0x1207fe0 (L)
    #  29-   228>	FUNC     	hf_select_ssl
    #  30-   240>	PUSH_P   	           0x12081e0 ("La")
    #  31-   250>	PUSH_U64 	                 0x0 (0)
    #  32-   260>	FUNC     	hf_accept_conn
    #  33-   278>	PUSH_U64 	                 0x0 (0)
    #  34-   288>	PUSH_P   	           0x1207fe0 (L)
    #  35-   298>	FUNC     	hf_select_ssl
    #  36-   2b0>	FUNC     	hf_accept_conn_none
    #  37-   2c8>	PUSH_U64 	                 0x0 (0)
    #  38-   2d8>	PUSH_P   	           0x12081e0 ("La")
    #  39-   2e8>	FUNC     	hf_select_ssl
    #  40-   300>	PUSH_P   	           0x12082a0
    #  41-   310>	PUSH_SIZE 	                 0x7 (7)
    #  42-   320>	FUNC     	hf_write
    #  43-   338>	PUSH_U64 	                 0x0 (0)
    #  44-   348>	PUSH_P   	           0x12080e0 (C)
    #  45-   358>	FUNC     	hf_select_ssl
    #  46-   370>	PUSH_P   	           0x12082a0
    #  47-   380>	PUSH_SIZE 	                 0x7 (7)
    #  48-   390>	FUNC     	hf_read_expect
    #  49-   3a8>	END      	
    #  50-   3b0>	+++      	
    # ----------------------------------------------------------------------
    # 
    # INFO:  @ test/radix/terp.c:525
    # executing script: script 'simple_conn' (simple connection to server)
    # 
    # --EXECUTION-----------------------------------------------------------
    #          0>	PUSH_P   	           0x1207fe0 (L)
    #         10>	PUSH_U64 	                 0x1 (1)
    #         20>	FUNC     	hf_new_ssl
    #         38>	PUSH_U64 	                 0x0 (0)
    #         48>	PUSH_P   	           0x1207fe0 (L)
    #         58>	FUNC     	hf_select_ssl
    #         70>	FUNC     	hf_listen
    #         88>	PUSH_P   	           0x12080e0 (C)
    #         98>	PUSH_U64 	                 0x0 (0)
    #         a8>	FUNC     	hf_new_ssl
    #         c0>	PUSH_U64 	                 0x0 (0)
    #         d0>	PUSH_P   	           0x12080e0 (C)
    #         e0>	FUNC     	hf_select_ssl
    #         f8>	PUSH_U64 	                 0x1 (1)
    #        108>	PUSH_P   	           0x1207fe0 (L)
    #        118>	FUNC     	hf_select_ssl
    #        130>	FUNC     	hf_set_peer_addr_from
    #        148>	PUSH_U64 	                 0x0 (0)
    #        158>	PUSH_P   	           0x12080e0 (C)
    #        168>	FUNC     	hf_select_ssl
    #        180>	FUNC     	hf_connect_wait
    #            		(span 1 times)
    #        198>	PUSH_U64 	                 0x0 (0)
    #        1a8>	PUSH_P   	           0x12080e0 (C)
    #        1b8>	FUNC     	hf_select_ssl
    #        1d0>	PUSH_P   	           0x1206a60
    #        1e0>	PUSH_SIZE 	                 0x6 (6)
    #        1f0>	FUNC     	hf_write
    #        208>	PUSH_U64 	                 0x0 (0)
    #        218>	PUSH_P   	           0x1207fe0 (L)
    #        228>	FUNC     	hf_select_ssl
    #        240>	PUSH_P   	           0x12081e0 ("La")
    #        250>	PUSH_U64 	                 0x0 (0)
    #        260>	FUNC     	hf_accept_conn
    #        278>	PUSH_U64 	                 0x0 (0)
    #        288>	PUSH_P   	           0x1207fe0 (L)
    #        298>	FUNC     	hf_select_ssl
    #        2b0>	FUNC     	hf_accept_conn_none
    #        2c8>	PUSH_U64 	                 0x0 (0)
    #        2d8>	PUSH_P   	           0x12081e0 ("La")
    #        2e8>	FUNC     	hf_select_ssl
    #        300>	PUSH_P   	           0x12082a0
    #        310>	PUSH_SIZE 	                 0x7 (7)
    #        320>	FUNC     	hf_write
    #        338>	PUSH_U64 	                 0x0 (0)
    #        348>	PUSH_P   	           0x12080e0 (C)
    #        358>	FUNC     	hf_select_ssl
    #        370>	PUSH_P   	           0x12082a0
    #        380>	PUSH_SIZE 	                 0x7 (7)
    #        390>	FUNC     	hf_read_expect
    #        3a8>	END      	
    # ----------------------------------------------------------------------
    # memory: 'Final state of stack'
    # 0000  empty
    #   (0 bytes)
    # 
    # Stats:
    #   Ops executed:               50
    # 
    # INFO:  @ test/radix/terp.c:525
    # completed: script 'simple_conn' (simple connection to server)
    # 
    # ======================================================================
    # Final process state for node 0, process 0:
    #   Threads (incl. main):        1
    #   Time slip:                   0 ms
    #   Objects:
    #       - L                @ 0x60d0000001e0
    #           Type:          QLSO
    #       - La               @ 0x614000004640
    #           Type:          QCSO
    #           Stream ID: 0
    #           Read state:    OK(1)
    #           Write state:   OK(1)
    #       - C                @ 0x614000001c40
    #           Type:          QCSO
    #           Stream ID: 0
    #           Read state:    OK(1)
    #           Write state:   OK(1)
    # 
    # ======================================================================
    # ==> OK
    # 
    ok 1 - iteration 1
    # INFO:  @ test/radix/terp.c:525
    # generating script: script 'simple_thread' (test that RADIX multithreading is working)
    # 
    # Generated script for 'simple_thread':
    # 
    # --GENERATED-----------------------------------------------------------
    #   # NAME:
    #   #   simple_thread
    #   # SOURCE:
    #   #   test/radix/quic_tests.c:32
    #   # DESCRIPTION:
    #   #   test that RADIX multithreading is working
    # 
    #   0-     0>	PUSH_P   	           0x1726e00
    #   1-    10>	FUNC     	hf_spawn_thread
    #   2-    28>	PUSH_P   	           0x1726e00
    #   3-    38>	FUNC     	hf_spawn_thread
    #   4-    50>	END      	
    #   5-    58>	+++      	
    # ----------------------------------------------------------------------
    # 
    # INFO:  @ test/radix/terp.c:525
    # executing script: script 'simple_thread' (test that RADIX multithreading is working)
    # 
    # --EXECUTION-----------------------------------------------------------
    #          0>	PUSH_P   	           0x1726e00
    #         10>	FUNC     	hf_spawn_thread
    #         28>	PUSH_P   	           0x1726e00
    #         38>	FUNC     	hf_spawn_thread
    #         50>	END      	
    # ----------------------------------------------------------------------
    # memory: 'Final state of stack'
    # 0000  empty
    #   (0 bytes)
    # 
    # Stats:
    #   Ops executed:                5
    # 
    # INFO:  @ test/radix/terp.c:525
    # completed: script 'simple_thread' (test that RADIX multithreading is working)
    # ==> Joining thread 1
    # ==> Joining thread 2
    # 
    # ====(n0/p0/t1)=======================================================
    # Result for child thread with index 1:
    #  -T- 1:	# INFO:  @ test/radix/terp.c:525
    #  -T- 1:	# generating script: script 'simple_thread_child' (test that RADIX multithreading is working (child))
    #  -T- 1:	# 
    #  -T- 1:	# Generated script for 'simple_thread_child':
    #  -T- 1:	# 
    #  -T- 1:	# --GENERATED-----------------------------------------------------------
    #  -T- 1:	#   # NAME:
    #  -T- 1:	#   #   simple_thread_child
    #  -T- 1:	#   # SOURCE:
    #  -T- 1:	#   #   test/radix/quic_tests.c:27
    #  -T- 1:	#   # DESCRIPTION:
    #  -T- 1:	#   #   test that RADIX multithreading is working (child)
    #  -T- 1:	# 
    #  -T- 1:	#   0-     0>	END      	
    #  -T- 1:	#   1-     8>	+++      	
    #  -T- 1:	# ----------------------------------------------------------------------
    #  -T- 1:	# 
    #  -T- 1:	# INFO:  @ test/radix/terp.c:525
    #  -T- 1:	# executing script: script 'simple_thread_child' (test that RADIX multithreading is working (child))
    #  -T- 1:	# 
    #  -T- 1:	# --EXECUTION-----------------------------------------------------------
    #  -T- 1:	#          0>	END      	
    #  -T- 1:	# ----------------------------------------------------------------------
    #  -T- 1:	# memory: 'Final state of stack'
    #  -T- 1:	#       NULL
    #  -T- 1:	#   (0 bytes)
    #  -T- 1:	# 
    #  -T- 1:	# Stats:
    #  -T- 1:	#   Ops executed:                1
    #  -T- 1:	# 
    #  -T- 1:	# INFO:  @ test/radix/terp.c:525
    #  -T- 1:	# completed: script 'simple_thread_child' (test that RADIX multithreading is working (child))
    #  -T- 1:	# 
    # ==> Child thread with index 1 exited with 1
    # 
    # ====(n0/p0/t2)=======================================================
    # Result for child thread with index 2:
    #  -T- 2:	# INFO:  @ test/radix/terp.c:525
    #  -T- 2:	# generating script: script 'simple_thread_child' (test that RADIX multithreading is working (child))
    #  -T- 2:	# 
    #  -T- 2:	# Generated script for 'simple_thread_child':
    #  -T- 2:	# 
    #  -T- 2:	# --GENERATED-----------------------------------------------------------
    #  -T- 2:	#   # NAME:
    #  -T- 2:	#   #   simple_thread_child
    #  -T- 2:	#   # SOURCE:
    #  -T- 2:	#   #   test/radix/quic_tests.c:27
    #  -T- 2:	#   # DESCRIPTION:
    #  -T- 2:	#   #   test that RADIX multithreading is working (child)
    #  -T- 2:	# 
    #  -T- 2:	#   0-     0>	END      	
    #  -T- 2:	#   1-     8>	+++      	
    #  -T- 2:	# ----------------------------------------------------------------------
    #  -T- 2:	# 
    #  -T- 2:	# INFO:  @ test/radix/terp.c:525
    #  -T- 2:	# executing script: script 'simple_thread_child' (test that RADIX multithreading is working (child))
    #  -T- 2:	# 
    #  -T- 2:	# --EXECUTION-----------------------------------------------------------
    #  -T- 2:	#          0>	END      	
    #  -T- 2:	# ----------------------------------------------------------------------
    #  -T- 2:	# memory: 'Final state of stack'
    #  -T- 2:	#       NULL
    #  -T- 2:	#   (0 bytes)
    #  -T- 2:	# 
    #  -T- 2:	# Stats:
    #  -T- 2:	#   Ops executed:                1
    #  -T- 2:	# 
    #  -T- 2:	# INFO:  @ test/radix/terp.c:525
    #  -T- 2:	# completed: script 'simple_thread_child' (test that RADIX multithreading is working (child))
    #  -T- 2:	# 
    # ==> Child thread with index 2 exited with 1
    # 
    # ======================================================================
    # Final process state for node 0, process 0:
    #   Threads (incl. main):        3
    #   Time slip:                   0 ms
    #   Objects:
    # 
    # ======================================================================
    # ==> OK
    # 
    ok 2 - iteration 2
ok 1 - test_script
```

Fixes openssl/project#448

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 30 days ago

hlandau commented:
Rebased

nhorman commented:
CI Looks relevant.  Do we typically allow the use of the PRI* formats to workaround the type formatting here, or is that not sufficiently portable?

nhorman commented on test/radix/quic_tests.c:
```diff
@@ -24,10 +24,25 @@ DEF_SCRIPT(simple_conn, "simple connection to server")
+    size_t i;
+
+    for (i = 0; i < 2; ++i)
+        OP_SPAWN_THREAD(simple_thread_child);
```
do you need a command macro here to join on the threads you create?  Or is that manageable in a future update?

hlandau replied:
This is done automatically by `quic_bindings.c` in `RADIX_PROCESS_join_all_threads`.

If we need to control joining explicitly in a test script, that can easily be added later. Just the framework for now.

hlandau commented:
Updated.

hlandau commented:
Worth noting &mdash; this is the only test currently in existence for the server API so it is a critical priority once the server API implementation PR is approved.

hlandau commented:
Rebased now the server API implementation PR is in.

Sashan commented on ssl/quic/quic_obj_local.h:
```diff
@@ -0,0 +1,280 @@
+    return obj->port;
+}
+
+/* Returns 1 iff this leader structure represents an event leader. */
```
c-style?

why ossl_unused? the function is used by `ossl_quic_obj_get0_engine_local()`

Sashan commented on ssl/quic/quic_obj_local.h:
```diff
@@ -0,0 +1,280 @@
+/*
+ * Similar to ossl_quic_obj_get0_engine, but only returns a non-NULL value if
+ * the obj object itself is an event leader, rather than one of its ancestors.
+ */
```
c-style?

Sashan commented on ssl/quic/quic_obj_local.h:
```diff
@@ -0,0 +1,280 @@
+        ? ossl_quic_obj_get0_engine(obj) : NULL;
+}
+
+/* Returns 1 iff this leader structure represents a port leader. */
```
c-style?

why ossl_unused, the function is called by  ossl_quic_obj_get0_port_local()

Sashan commented on ssl/quic/quic_obj_local.h:
```diff
@@ -0,0 +1,280 @@
+ * Similar to ossl_quic_obj_get0_port, but only returns a non-NULL value if
+ * the obj object itself is a port leader, rather than one of its ancestors.
+ */
+static ossl_inline ossl_unused QUIC_PORT *
```
c-style?

Sashan commented on ssl/quic/quic_obj_local.h:
```diff
@@ -0,0 +1,280 @@
+/*
+ * Convenience Inlines
+ * ===================
+ */
```
I wonder if the _section tile_ should be moved to higher up, it does not make much sense in the middle of various _inline_ definitions here.

Sashan commented on ssl/quic/quic_obj_local.h:
```diff
@@ -0,0 +1,280 @@
+ */
+
+/* Get a pointer to the QUIC domain mutex. Always returns non-NULL. */
+static ossl_inline ossl_unused CRYPTO_MUTEX *
```
c-style?

function is actually used here:  373  ssl/quic/quic_impl.c <<qctx_lock>>

Sashan commented on ssl/quic/quic_obj_local.h:
```diff
@@ -0,0 +1,280 @@
+ * non-NULL.
+ */
+static ossl_inline ossl_unused QUIC_REACTOR *
+ossl_quic_obj_get0_reactor(const QUIC_OBJ *obj)
```
c-style?

function is actually being used here: ossl_quic_handle_events() in ssl/quic/quic_impl.c

Sashan commented on ssl/quic/quic_obj_local.h:
```diff
@@ -0,0 +1,280 @@
+}
+
+/* Get a reference to the OSSL_LIB_CTX pointer applicable to a leader. */
+static ossl_inline ossl_unused OSSL_LIB_CTX *
```
c-style?

Sashan commented on ssl/quic/quic_obj_local.h:
```diff
@@ -0,0 +1,280 @@
+}
+
+/* Get a reference to the propq pointer applicable to a leader. */
+static ossl_inline ossl_unused const char *
```
c-style?

Sashan commented on ssl/quic/quic_obj_local.h:
```diff
@@ -0,0 +1,280 @@
+ * Returns the APL object pointer to the event leader in a hierarchy. Always
+ * returns non-NULL.
+ */
+static ossl_inline ossl_unused SSL *
```
c-style?

Sashan commented on ssl/quic/quic_obj_local.h:
```diff
@@ -0,0 +1,280 @@
+ * Returns the APL object pointer to the port leader in a hierarchy (if any).
+ * Always returns non-NULL.
+ */
+static ossl_inline ossl_unused SSL *
```
c-style?

Sashan commented on ssl/quic/quic_port.c:
```diff
@@ -244,6 +251,65 @@ int ossl_quic_port_update_poll_descriptors(QUIC_PORT *port)
+ * Our strategy for determining address mode is simple: we probe the underlying
+ * network BIOs for their capabilities. If the network BIOs support what we
+ * need, we use addressed mode. Otherwise, we use unaddressed mode.
+ *
```
I think the last paragraph lines 279 - 281 should be deleted. Perhaps other parts of comment here needs to be revisited because it came from different location. So it needs to be adjusted to match new context here.

hlandau replied:
This doesn't seem out of place to me...

Sashan replied:
My confusion comes from sentence found in the last paragraph:
```
 * If addressed mode is chosen, we require an initial peer address to be set. If
 * this is not set, we fail. If unaddressed mode is used, we do not require
 * this, as such an address is superfluous, though it can be set if desired.
```
`If this is not set, we fail.` -- However the function `port_update_addressing_mode()` is a void function returns nothing. So this made me wonder how we indicate the failure to caller. So this is what makes me thinking the comment needs update after being moved here.  It was just question to raise awareness.

Sashan commented on ssl/ssl_lib.c:
```diff
@@ -7478,6 +7475,23 @@ int SSL_is_connection(SSL *s)
+{
+#ifndef OPENSSL_NO_QUIC
+    if (!IS_QUIC(s))
+        return s;
```
why returning `s` at line 7482 is correct? The reason I ask is that for non-QUIC case we just assume every ssl object is listener. I'm not sure it's desired.

Sashan commented on ssl/ssl_local.h:
```diff
@@ -1810,38 +1814,6 @@ struct ssl_connection_st {
-    SSL_CONNECTION_FROM_SSL_ONLY_int(ssl, const)
-# endif
-
 /*
```
Just a comment suggestion. IMO it would be better to put stuff above (moving macros to ssl_unwrap.h) to dedicated PR next time.

Sashan commented on ssl/t1_trce.c:
```diff
@@ -13,6 +13,10 @@
 /* Packet trace support for OpenSSL */
 #include "internal/nelem.h"
+#include "internal/ssl_unwrap.h"
+#include "internal/quic_predef.h"
```
I think we don't need to include quic_predef.h

Sashan commented on ssl/t1_trce.c:
```diff
@@ -13,6 +13,10 @@
+#include "internal/ssl_unwrap.h"
+#include "internal/quic_predef.h"
+#include "internal/quic_trace.h"
+#include "quic/quic_local.h"
```
I think we don't need quic_local.h
removing those two headers does not break compilation at least in my local workspace.

Sashan commented on test/radix/README.md:
```diff
@@ -0,0 +1,114 @@
+- multi-process (in future)
+- multi-node (in future)
+
+test vignettes for network protocol testing applications. While it is currently
```
question: what are test vignettes? are they kind of set of tests? or is it a different name for a script? or test scenario?

hlandau replied:
Yes &mdash; basically a scenario. An occasionally used term...

Sashan commented on test/radix/quic_bindings.c:
```diff
@@ -0,0 +1,788 @@
+    if (!TEST_ptr(obj = OPENSSL_zalloc(sizeof(*obj))))
+       return NULL;
+
+    obj->name = OPENSSL_strdup(name);
```
no `if (!TEST_PTR(obj->name = OPENSSL_strdup(name))` check here? is that intentional?

Sashan commented on test/radix/quic_ops.c:
```diff
@@ -0,0 +1,1055 @@
+        goto err;
+
+    if (ret != 1) {
+        if (1 /* TODO */ && is_want(ssl, ret))
```
TODO what? I would prefer to have at least decent hint to see what's missing here.

Sashan reviewed (changes requested):
many of my comments are nits feel free to ignore them.

I have a question: while working on anti-amplification code, I spent some time in `test/quic_fc_test.c` is this test going to be ported to RADIX framework too? I'm still trying to figure out how to extend the test to cover my amplification changes. I hope to resume to work on test on Friday hopefully.

hlandau commented on doc/designs/quic-design/server/quic-polling.md:
```diff
@@ -0,0 +1,1658 @@
+     * disable_events and enable_events manages registered event types.
+     * This is used to disable and enable OSSL_POLL_EVENT_FLAG bits.
+     */
+    uint64_t            disable_flags, enable_flags;
```
There will be more flags over time, yes. We have been using `uint64_t` for flags arguments for a while now for this reason.

hlandau commented on ssl/quic/quic_local.h:
```diff
@@ -119,18 +120,25 @@ struct quic_xso_st {
 
+    /* The QLSO this connection belongs to, if any. */
+    QUIC_LISTENER                   *listener;
+
```
A `quic_conn_st` is never a listener. But it may have been created by a listener, in which case `listener` is non-NULL. If a `quic_conn_st` (`QUIC_CONNECTION`) was created directly then `listener` is NULL, yes.

hlandau commented on ssl/quic/quic_obj_local.h:
```diff
@@ -0,0 +1,280 @@
+ * In the libssl APL, we have QLSOs, QCSOs and QSSOs, and in the future might
+ * choose to introduce QDSOs. There are also roles such as Port Leader and Event
+ * Leader which can be assumed by these different types under different
+ * circumstances — in other words, whether an APL object is a Port or Event
```
I don't believe we have any rule against this...

hlandau commented on ssl/quic/quic_obj_local.h:
```diff
@@ -0,0 +1,280 @@
+ *
+ * This structure must come at the start of a QUIC object structure definition.
+ *
+ * ssl->type still determines the actual object type. An SSL object pointer s
```
This is intentional

hlandau commented on ssl/quic/quic_obj_local.h:
```diff
@@ -0,0 +1,280 @@
+ * NULL. If obj is non-NULL, it must be initialised.
+ */
+static ossl_inline ossl_unused SSL *
+ossl_quic_obj_get0_ssl(QUIC_OBJ *obj)
```
It is necessary to break many of these over multiple lines to keep the lines within length... Or is your comment related to where a `*` is placed?

Sashan replied:
in [c-style](https://www.openssl.org/policies/technical/coding-style.html) we often show function definition like this
```
int function(int x)
{
        body of function
}
```
So I've just jumped to wrong assumption. In fact what you have currently is style I prefer more. Because I can quickly find function definition by just using search pattern `/^func_name`.

the c-style guide does not talk at all on how functions should be defined.

hlandau commented on ssl/quic/quic_obj_local.h:
```diff
@@ -0,0 +1,280 @@
+/*
+ * Determines the applicable engine and return a pointer to it. Never returns
+ * NULL.
+ */
```
We habitually tag inlines with `ossl_unused` so that they do not cause warnings if they do not happen to be used. It is basically done throughout the codebase whenever we use `ossl_inline`.

hlandau commented on ssl/quic/quic_obj_local.h:
```diff
@@ -0,0 +1,280 @@
+    return obj->engine;
+}
+
+/* Determines the applicable port (if any) and returns a pointer to it. */
```
See above

hlandau commented:
Rebased and updated now that the server API implementation PR is in.

>I have a question: while working on anti-amplification code, I spent some time in test/quic_fc_test.c is this test going to be ported to RADIX framework too? I'm still trying to figure out how to extend the test to cover my amplification changes. I hope to resume to work on test on Friday hopefully.

There are no plans to do that at this stage.

Note that the FC code should not have anything to do with amplification changes. The FC code is for stream-level flow control. Amplification limits have to be applied at the datagram level.

Sashan reviewed (approved):
I gave one more quick look to the code. I think this is good enough to go in. So more people can start using and improving it.

nhorman commented:
changes look good, my approval holds

t8m commented on test/radix/quic_bindings.c:
```diff
@@ -0,0 +1,768 @@
+    case SSL_STREAM_STATE_OK:
+        return "OK";
+    case SSL_STREAM_STATE_WRONG_DIR:
+        return "wrong dir";
```
In cases below you use dash between words - should it be `wrong-dir`?

t8m reviewed (changes requested):
Great work!

Just one minor nit.

hlandau commented:
Updated.

t8m reviewed (approved):
@nhorman assuming still approving

nhorman commented:
@hlandau yes, approval holds

Sashan reviewed (approved):
still looks good.

openssl-machine commented:
24 hours has passed since 'approval: done' was set, but as this PR has been updated in that time the label 'approval: ready to merge' is not being automatically set.  Please review the updates and set the label manually.

hlandau commented:
Merged to feature branch. Thanks for the reviews.
